This course covers the key Python skills you’ll need so you can start using Python for data science. 

We'll start with a brief overview of Python syntax, variable assignment, and arithmetic operators.

# Lesson 1 - Hello Python!

## Lesson:

Python was named for the British comedy troupe [Monty Python](https://en.wikipedia.org/wiki/Monty_Python).

Just for fun, try reading over the code below and predicting what it's going to do when run. (If you have no idea, that's fine!)

Then click the "output" button to see the results of our program.

In [1]:
lesson_amount = 0
print(lesson_amount)

#We learned about Python in class, so let add 1 into our lesson_amount
lesson_amount = lesson_amount + 1

#Let's check if we have learn some lesson of Python code before
if lesson_amount > 0:
    print("I learn some Python code!")

lesson_learned = "Lesson " * lesson_amount
print(lesson_learned)

0
I learn some Python code!
Lesson 


There's a lot to unpack here! This silly program demonstrates many important aspects of what Python code looks like and how it works. Let's review the code from top to bottom.

In [2]:
lesson_amount = 0

**Variable assignment:** Here we create a variable called `lesson_amount` and assign it the value of 0 using `=`, which is called the assignment operator.

> **Aside**: If you've programmed in certain other languages (like Java or C++), you might be noticing some things Python *doesn't* require us to do here:  
- we don't need to "declare" `lesson_amount` before assigning to it
- we don't need to tell Python what type of value `lesson_amount` is going to refer to. In fact, we can even go on to reassign `lesson_amount` to refer to a different sort of thing like a string or a boolean.

In [3]:
print(lesson_amount)

0


**Function calls:**. `print` is a Python function that displays the value passed to it on the screen. We call functions by putting parentheses after their name, and putting the inputs (or *arguments*) to the function in those parentheses.

In [4]:
#We learned about Python in class, so let add 1 into our lesson_amount
lesson_amount = lesson_amount + 1

The first line above is a **comment**. In Python, comments begin with the `#` symbol.

Next we see an example of reassignment. Reassigning the value of an existing variable looks just the same as creating a variable - it still uses the `=` assignment operator.

In this case, the value we're assigning to `lesson_amount` involves some simple arithmetic on its previous value. When it encounters this line, Python evaluates the expression on the right-hand-side of the `=` (0 + 1 = 1), and then assigns that value to the variable on the left-hand-side.

In [5]:
#Let's check if we have learn some lesson of Python code before
if lesson_amount > 0:
    print("I learn some Python code!")

lesson_learned = 'Lesson Lesson Lesson'
print(lesson_learned)

I learn some Python code!
Lesson Lesson Lesson


We won't talk much about "conditionals" until later, but, even if you've never coded before, you can probably guess what this does. Python is prized for its readability and the simplicity. 

Note how we indicated which code belongs to the `if`. `"I learned some Python code!"` is only supposed to be printed if `lesson_amount` is positive. But the later code (like `print(lesson_learned)`) should be executed no matter what. How do we (and Python) know that?

The colon (`:`) at the end of the `if` line indicates that a new **code block** is starting. Subsequent lines which are indented are part of that code block. 
> **Aside**: If you've coded before, you might know that some other languages use `{`curly braces`}` to mark the beginning and end of code blocks. Python's use of meaningful whitespace can be surprising to programmers who are accustomed to other languages, but in practice it can lead to more consistent and readable code than languages that do not enforce indentation of code blocks. 

The later lines dealing with `lesson_learned` are not indented with an extra 4 spaces, so they're not a part of the `if`'s code block. We'll see more examples of indented code blocks later when we define functions and using loops.

This code snippet is also our first sighting of a **string** in Python:

```python
"I learned some Python code!"
```

Strings can be marked either by double or single quotation marks. (But because this particular string *contains* a single-quote character, we might confuse Python by trying to surround it with single-quotes, unless we're careful.)

In [6]:
lesson_learned = "Lesson " * lesson_amount
print(lesson_learned)

Lesson 


The `*` operator can be used to multiply two numbers (`3 * 3` evaluates to 9), but we can also multiply a string by a number, to get a version that's been repeated that many times. Python offers a number of cheeky little time-saving tricks like this where operators like `*` and `+` have a different meaning depending on what kind of thing they're applied to. (The technical term for this is [operator overloading](https://en.wikipedia.org/wiki/Operator_overloading).)

## Numbers and arithmetic in Python

We've already seen an example of a variable containing a number above:

In [7]:
lesson_amount = 0

"Number" is a fine informal name for the kind of thing, but if we wanted to be more technical, we could ask Python how it would describe the type of thing that `lesson_amount` is:

In [8]:
type(lesson_amount)

int

It's an `int` - short for integer. There's another sort of number we commonly encounter in Python:

In [9]:
type(19.95)

float

A `float` is a number with a decimal place - very useful for representing things like weights or proportions.

`type()` is the second built-in function we've seen (after `print()`), and it's another good one to remember. It's very useful to be able to ask Python "what kind of thing is this?". 

A natural thing to want to do with numbers is perform arithmetic. We've seen the `+` operator for addition, and the `*` operator for multiplication. Python also has us covered for the rest of the basic buttons on your calculator:

| Operator     | Name           | Description                                            |
|--------------|----------------|--------------------------------------------------------|
| ``a + b``    | Addition       | Sum of ``a`` and ``b``                                 |
| ``a - b``    | Subtraction    | Difference of ``a`` and ``b``                          |
| ``a * b``    | Multiplication | Product of ``a`` and ``b``                             |
| ``a / b``    | True division  | Quotient of ``a`` and ``b``                            |
| ``a // b``   | Floor division | Quotient of ``a`` and ``b``, removing fractional parts |
| ``a % b``    | Modulus        | Integer remainder after division of ``a`` by ``b``     |
| ``a ** b``   | Exponentiation | ``a`` raised to the power of ``b``                     |
| ``-a``       | Negation       | The negative of ``a``                                  |

<span style="display:none"></span>

One interesting observation here is that, whereas your calculator probably just has one button for division, Python can do two kinds. "True division" is basically what your calculator does:

In [10]:
print(5 / 2)
print(6 / 2)

2.5
3.0


It always gives us a `float`. 

The `//` operator gives us a result that's rounded down to the next integer.

In [11]:
2 + 2

4

In [12]:
print(5 // 2)
print(6 // 2)

2
3


Can you think of where this would be useful? You'll see an example soon in the coding challenges.

### Order of operations

The arithmetic we learned in primary school has conventions about the order in which operations are evaluated. Some remember these by a mnemonic such as **PEMDAS** - **P**arentheses, **E**xponents, **M**ultiplication/**D**ivision, **A**ddition/**S**ubtraction.

Python follows similar rules about which calculations to perform first. They're mostly pretty intuitive.

In [13]:
8 - 3 + 2

7

In [14]:
-3 + 4 * 2

5

Sometimes the default order of operations isn't what we want:

In [15]:
hat_height_cm = 25
my_height_cm = 190
# How tall am I, in meters, when wearing my hat?
total_height_meters = hat_height_cm + my_height_cm / 100
print("Height in meters =", total_height_meters, "?")

Height in meters = 26.9 ?


Parentheses are useful here. You can add them to force Python to evaluate sub-expressions in whatever order you want.

In [16]:
total_height_meters = (hat_height_cm + my_height_cm) / 100
print("Height in meters =", total_height_meters)

Height in meters = 2.15


### Builtin functions for working with numbers

`min` and `max` return the minimum and maximum of their arguments, respectively...

In [17]:
print(min(1, 2, 3))
print(max(1, 2, 3))

1
3


`abs` returns the absolute value of an argument:

In [18]:
print(abs(32))
print(abs(-32))

32
32


In addition to being the names of Python's two main numerical types, `int` and `float` can also be called as functions which convert their arguments to the corresponding type:

In [19]:
print(float(10))
print(int(3.33))
# They can even be called on strings!
print(int('807') + 1)

10.0
3
808


## Exercise:

Notebooks are composed of blocks (called "cells") of text and code. Each of these is editable, though you'll mainly be editing the code cells to answer some questions.

To get started, try running the code cell below (by pressing the ► button, or clicking on the cell and pressing ctrl+enter on your keyboard).

In [20]:
print('Hello, World!')

Hello, World!


In [21]:
print("You've successfully run some Python code")
print("Congratulations!")

You've successfully run some Python code
Congratulations!


Try adding another line of code in the cell above and re-running it. 

Now let's get a little fancier:  Add a new code cell by clicking on an existing code cell, hitting the escape key, and then hitting the `a` or `b` key.  The `a` key will add a cell above the current cell, and `b` adds a cell below.

Great! Now you know how to use Notebooks.

Each hands-on exercise starts by setting up our feedback and code checking mechanism. Run the code cell below to do that. Then you'll be ready to move on to question 0.

### 0.

*This is a silly question intended as an introduction to the format we use for hands-on exercises throughout all courses.*

**What is your favorite color? **

In [23]:
# create a variable called color with an appropriate value on the line below
# (Remember, strings in Python must be enclosed in 'single' or "double" quotes)
color = 'Red'
print(color)

Red


<hr/>

# 1.

Complete the code below. In case it's helpful, here is the table of available arithmetic operations:



| Operator     | Name           | Description                                            |
|--------------|----------------|--------------------------------------------------------|
| ``a + b``    | Addition       | Sum of ``a`` and ``b``                                 |
| ``a - b``    | Subtraction    | Difference of ``a`` and ``b``                          |
| ``a * b``    | Multiplication | Product of ``a`` and ``b``                             |
| ``a / b``    | True division  | Quotient of ``a`` and ``b``                            |
| ``a // b``   | Floor division | Quotient of ``a`` and ``b``, removing fractional parts |
| ``a % b``    | Modulus        | Integer remainder after division of ``a`` by ``b``     |
| ``a ** b``   | Exponentiation | ``a`` raised to the power of ``b``                     |
| ``-a``       | Negation       | The negative of ``a``                                  |

<span style="display:none"></span>


In [25]:
pi = 3.14159 # approximate
diameter = 3

# Create a variable called 'radius' equal to half the diameter
radius = diameter/2

# Create a variable called 'area', using the formula for the area of a circle: pi times the radius squared
area = pi * (radius**2)

# Print the result of area
print(area)

7.0685775


Result: area = 7.0685775

<hr/>

# 2.

Add code to the following cell to swap variables `a` and `b` (so that `a` refers to the object previously referred to by `b` and vice versa).

In [26]:
########### Setup code - don't touch this part ######################
# If you're curious, these are examples of lists. We'll talk about 
# them in depth a few lessons from now. For now, just know that they're
# yet another type of Python object, like int or float.
a = [1, 2, 3]
b = [3, 2, 1]
######################################################################

# Your code goes here. Swap the values to which a and b refer.
# Hint: Try using a third variable
c = a
a = b
b = c
print(a)
print(b)

[3, 2, 1]
[1, 2, 3]


<hr/>

# 3a.

Add parentheses to the following expression so that it evaluates to 1.

In [27]:
(5 - 3) // 2

1

# 3b.  <span title="A bit spicy" style="color: darkgreen ">🌶️</span>

<small>Questions, like this one, marked a spicy pepper are a bit harder.</small>

Add parentheses to the following expression so that it evaluates to 0.

In [28]:
8 - (3 * 2) - (1 + 1)

0

<hr/>

# 4. 
Alice, Bob and Carol have agreed to pool their Halloween candy and split it evenly among themselves.
For the sake of their friendship, any candies left over will be smashed. For example, if they collectively
bring home 91 candies, they'll take 30 each and smash 1.

Write an arithmetic expression below to calculate how many candies they must smash for a given haul.

Hint: You'll probably want to use the modulo operator, %, to obtain the remainder of division.

In [30]:
# Variables representing the number of candies collected by alice, bob, and carol
alice_candies = 121
bob_candies = 77
carol_candies = 109

# Your code goes here! Replace the right-hand side of this assignment with an expression
# involving alice_candies, bob_candies, and carol_candies
to_smash = (alice_candies + bob_candies + carol_candies) % 3
print(to_smash)


1
